File: code_with_inputs/calculate_probability.py
def calculate_probability(dc, bonus_min=0, bonus_max=0, advantage=False, dice_sides=20):
    calcfailprob = lambda b: max(1, dc - b - 1) / dice_sides
    factor = {True: 2, False: 1}.get(advantage, 1)
    prob_min = max(1 - calcfailprob(bonus_min) ** factor, 0)
    prob_max = max(1 - calcfailprob(bonus_max) ** factor, 0)
    return {"min": prob_min, "max": prob_max, "total": (prob_max + prob_min) / 2}


if __name__ == "__main__":
    dc = int(input("Enter the Difficulty Class (DC): "))
    bonus_min = int(input("Enter the minimum bonus: "))
    bonus_max = int(input("Enter the maximum bonus: "))
    advantage = int(input("Is there an advantage? (0 for no, 1 for yes): "))
    dice_sides = 20

    probability = calculate_probability(dc, bonus_min, bonus_max, advantage, dice_sides)
    print(probability)

File: code_with_inputs/calculate_probability.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <algorithm>
#include <numeric>
#include <cmath>

struct ProbabilityResult {
    double min;
    double max;
    double total;
};

ProbabilityResult calculateProbability(int dc, int bonusMin, int bonusMax, bool advantage, int diceSides) {
    double factor = advantage ? 2 : 1;
    double failProbMin = std::max(1, dc - bonusMin - 1) / static_cast<double>(diceSides);
    double failProbMax = std::max(1, dc - bonusMax - 1) / static_cast<double>(diceSides);

    double probMin = std::max(1 - std::pow(failProbMin, factor), 0.0);
    double probMax = std::max(1 - std::pow(failProbMax, factor), 0.0);
    double total = (probMax + probMin) / 2;

    return {probMin, probMax, total};
}


int main() {
    int dc;
    int bonusMin;
    int bonusMax;
    bool advantage;
    const int diceSides = 20;

    std::cout << "Enter the Difficulty Class (DC): ";
    std::cin >> dc;
    std::cout << "Enter the minimum bonus: ";
    std::cin >> bonusMin;
    std::cout << "Enter the maximum bonus: ";
    std::cin >> bonusMax;
    std::cout << "Is there an advantage? (0 for no, 1 for yes): ";
    std::cin >> advantage;

    ProbabilityResult probabilities = calculateProbability(dc, bonusMin, bonusMax, advantage, diceSides);
    
    std::cout << "Probability Min: " << probabilities.min << std::endl;
    std::cout << "Probability Max: " << probabilities.max << std::endl;
    std::cout << "Total Probability: " << probabilities.total << std::endl;

    return 0;
}